### ISSUE TITLE
[Storybook 6.4.beta-7]  composeStories causes Types of property 'default' are incompatible with CSF3.0

### ISSUE THREAD

**Issue by the-ult** [storybookjs/testing-angular#1]
**Describe the bug**
Using storybook 6.4.beta-7 with CSF3.0 seems not to be compatible.

```ts
const { Default, withTitleElement } = composeStories(stories);
```
causes
```
Argument of type 'typeof import(".../app-bar.component.stories")' is not assignable to parameter of type '{ default: Meta<Args>; __esModule?: boolean | undefined; }'.
  Types of property 'default' are incompatible.
    Type 'import(".../node_modules/@storybook/angular/dist/ts3.9/client/preview/types-6-0").Meta<import(".../node_modules/@storybook/csf/dist/story").Args>' is not assignable to type 'import(".../node_modules/@storybook/testing-angular/node_modules/@storybook/angular/dist/ts3.9/client/preview/types-6-0").Meta<import(".../node_modules/@storybook/testing-angular/node_modules/@storybook/addons/dist/ts3.9/types").Args>'.
      Type 'Meta<Args>' is not assignable to type 'BaseMeta<any>'.
        Types of property 'title' are incompatible.
          Type 'string | undefined' is not assignable to type 'string'.
            Type 'undefined' is not assignable to type 'string'.ts(2345)
```

**To Reproduce**
 example-code
`app-bar.component.stories.ts`
```ts
import { CommonModule } from '@angular/common';
import { moduleMetadata } from '@storybook/angular';

import { AppBarTitleComponent } from './app-bar-title.component';
import { AppBarComponent } from './app-bar.component';

export default {
  title: 'app-bar',
  component: AppBarComponent,
  subcomponents: { AppBarTitleComponent },
  decorators: [
    moduleMetadata({
      imports: [CommonModule],
      declarations: [AppBarTitleComponent],
    }),
  ],
};

export const Default = {};

export const withTitleElement = {
  render: (props: AppBarComponent) => ({
    props,
    template: `
    <ddm-app-bar>
      <ddm-app-bar-title>MY-TITLE</ddm-app-bar-title>
    </ddm-app-bar>
    `,
  }),
};

... 
```

```ts
import { composeStories, createMountableStoryComponent } from '@storybook/testing-angular';
import { render, screen } from '@testing-library/angular';

import * as stories from './app-bar.component.stories'; // import all stories from the stories file

// Every component that is returned maps 1:1 with the stories, but they already contain all decorators from story level,
// meta level and global level.
const { Default, withTitleDirective } = composeStories(stories);

describe('AppBarComponent', () => {
  it('should create the default AppBar', async () => {
    const { component, ngModule } = createMountableStoryComponent(Default({}, {} as unknown));

    await render(component, { imports: [ngModule] });

    const brand = screen.getByTitle('brand');
    expect(brand).toBeInTheDocument();
  });
}
```

`main.js
```js
module.exports = {
  stories: [],
  addons: [
    // '@storybook/angular', // <- disabled because of: https://github.com/storybookjs/storybook/issues/16084#issuecomment-927722256
    '@storybook/addon-essentials',
    {
      name: '@storybook/addon-storysource',
      options: {
        loaderOptions: {
          injectStoryParameters: false,
        },
      },
    },
    '@storybook/addon-links',
  ],
};

```

**Additional context**
```
❯ npx sb@next info                             
Environment Info:

  System:
    OS: macOS 11.6
    CPU: (12) x64 Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
  Binaries:
    Node: 14.18.0 - ~/.nvm/versions/node/v14.18.0/bin/node
    npm: 7.24.1 - ~/.nvm/versions/node/v14.18.0/bin/npm
  Browsers:
    Chrome: 94.0.4606.71
    Edge: 94.0.992.38
    Safari: 15.0
  npmPackages:
    @storybook/addon-actions: 6.4.0-beta.7 => 6.4.0-beta.7 
    @storybook/addon-essentials: 6.4.0-beta.7 => 6.4.0-beta.7 
    @storybook/addon-jest: 6.4.0-beta.7 => 6.4.0-beta.7 
    @storybook/addon-links: 6.4.0-beta.7 => 6.4.0-beta.7 
    @storybook/addon-storysource: 6.4.0-beta.7 => 6.4.0-beta.7 
    @storybook/angular: 6.4.0-beta.7 => 6.4.0-beta.7 
    @storybook/builder-webpack5: 6.4.0-beta.7 => 6.4.0-beta.7 
    @storybook/manager-webpack5: 6.4.0-beta.7 => 6.4.0-beta.7 
    @storybook/testing-angular: 0.0.9 => 0.0.9 
```
```
❯ nx report

>  NX  Report complete - copy this into the issue template

  Node : 14.18.0
  OS   : darwin x64
  npm  : 7.24.1
  
  nx : 12.10.0
  @nrwl/angular : 12.10.0
  @nrwl/cli : 12.10.0
  @nrwl/cypress : 12.10.0
  @nrwl/devkit : 12.10.0
  @nrwl/eslint-plugin-nx : 12.10.0
  @nrwl/jest : 12.10.0
  @nrwl/linter : 12.10.0
  @nrwl/tao : 12.10.0
  @nrwl/workspace : 12.10.0
  @nrwl/storybook : 12.10.0
  typescript : 4.3.5
```

Might be related to: https://github.com/storybookjs/storybook/issues/16084 ??

**Comment by the-ult** [comment]
Don't know if it is related or a separate issue, but running the jest test results in:

```
 FAIL  .../app-bar/app-bar.component.spec.ts
  ● Test suite failed to run

    ReferenceError: EventSource is not defined

      1 | import { CommonModule } from '@angular/common';
    > 2 | import { moduleMetadata } from '@storybook/angular';
        | ^
      3 |
      4 | import { AppBarTitleComponent } from './app-bar-title.component';
      5 | import { AppBarComponent } from './app-bar.component';

      at Object.<anonymous> (../../../node_modules/@storybook/preview-web/dist/cjs/StoryIndexClient.js:120:34)
      at Object.<anonymous> (../../../node_modules/@storybook/preview-web/dist/cjs/PreviewWeb.js:58:25)
      at Object.<anonymous> (../../../node_modules/@storybook/preview-web/dist/cjs/index.js:41:19)
      at Object.<anonymous> (../../../node_modules/@storybook/core-client/dist/cjs/preview/start.js:38:19)
      at Object.<anonymous> (../../../node_modules/@storybook/core-client/dist/cjs/preview/index.js:38:14)
      at Object.<anonymous> (../../../node_modules/@storybook/core-client/dist/cjs/index.js:11:16)
      at Object.<anonymous> (../../../node_modules/@storybook/core/dist/cjs/index.js:7:19)
      at Object.<anonymous> (../../../node_modules/@storybook/core/client.js:1:100)
      at Object.<anonymous> (../../../node_modules/@storybook/angular/dist/ts3.9/client/preview/index.js:14:16)
      at Object.<anonymous> (../../../node_modules/@storybook/angular/dist/ts3.9/client/index.js:13:17)
      at Object.<anonymous> (src/lib/components/app-bar/app-bar.component.stories.ts:2:1)
      at Object.<anonymous> (src/lib/components/app-bar/app-bar.component.spec.ts:4:1)
      at TestScheduler.scheduleTests (../../../node_modules/@jest/core/build/TestScheduler.js:333:13)
      at runJest (../../../node_modules/@jest/core/build/runJest.js:387:19)
      at _run10000 (../../../node_modules/@jest/core/build/cli/index.js:408:7)
      at runCLI (../../../node_modules/@jest/core/build/cli/index.js:261:3)
```

**Comment by Marklb** [comment]
If you use `6.4.x` there will be types errors. Later, I will see if I can get a beta version published with the types working for `6.4.x`. The types were fine if I compiled this with a `6.4.x` version, but I didn't try CSF3. I am not positive if it would fix the types for CSF3, but I am importing most of the types from Storybook, so hopefully it will work for both CSF versions. Using a CSF3 story, it seemed to render fine, but I only did it once to see what would happen. I haven't actually tried switching my projects stories to it, yet.

I published with the types working for `6.3.8`, because I wasn't able to figure out a way to build a version with types that worked for both versions.

I have been using Storybook `6.4.0-beta.7` in one of my projects, but I had to do `const { Default } = composeStories(stories as any)`. Even though it compiles and works fine it does ruin the types.

The "EventSource is not defined" error is not something I had seen before, because I haven't been using Jest for my Story tests. I just tried in my project that uses Jest and got the same error, as well as a compodoc export error when setting the global config from `.storybook/preview.ts`, but neither should be related to this issue.

For the EventSource error, I took a look at the recent changes in Storybook and see that they avoided the error with the following, which also worked for my project: https://github.com/storybookjs/storybook/blob/next/app/angular/setup-jest.ts#L4

**Comment by the-ult** [comment]
Thanks for the `EventSource` tip.

Getting a new error now. But gonna dive into that one tomorrow.

```ts
TypeError: story is not a function

      13 | describe('AppBarComponent', () => {
      14 |   it('should create the default AppBar', async () => {
    > 15 |     const { component, ngModule } = createMountableStoryComponent(Default({}, {} as any));
         |                                                                   ^
      16 |
      17 |     await render(component, { imports: [ngModule] });
```

**Comment by Marklb** [comment]
Yeah, that is another type issue that I need to look at. I was trying to wait on `6.4.x` to get farther along before I spent too much time dealing with the rest of the types, but now that it's in beta they probably won't change much, so I should get them fixed.

I have been playing with ways to reduce the boilerplate of returning component/module and passing them directly into whatever is rendering them. The following is what I used for `@testing-library/angular`. It of course doesn't improve the types, as it is, but all my stories so far haven't needed to pass anything else to the render function. Also, it gave me less places to refactor when the types are fixed, since I don't need to add `as any` to each StoryFn and instead I should be able to just update my types on the following function's arguments.
```ts
import { createMountableStoryComponent } from '@storybook/testing-angular'
import { render, RenderResult } from '@testing-library/angular'

export async function renderStory<ComponentType = any>(story: any, props: any = {}): Promise<RenderResult<ComponentType, ComponentType>> {
  const { component, ngModule } = createMountableStoryComponent(story({}, props))
  return await render(component, { imports: [ ngModule ] })
}

// Then I can just do the following for the test:
it('should ...', async () => {
  renderStory(MyStory)
  ...
  expect(...).toBe(...)
})
```

Only problem, with a `renderStory` function is that it's implementation would be unique per library. So, I am thinking it may not reduce enough boilerplate to be worth maintaining them. Cypress has a `mountStory` function in an incomplete Angular implementation of their component testing, but I don't know if it is better for the testing libraries to decide if they want to add a function like that, this library provide a common `renderStory` function type that wraps the commonly used testing libraries, or neither and just leave it as is with the types fixed.

**Comment by the-ult** [comment]
Hey @Marklb, 

Thanks for the explanation and sample code.
Somehow I keep getting the same error: (but in this case within the RenderStory function)
```
● AppBarComponent › should create the default AppBar

    TypeError: story is not a function

       9 |   props: any = {},
      10 | ): Promise<RenderResult<ComponentType, ComponentType>> {
    > 11 |   const { component, ngModule } = createMountableStoryComponent(story({}, props));
         |                                                                 ^
      12 |   return await render(component, { imports: [ngModule] });
      13 | }
```

For now, I'm gonna write the test the testing-library way. (Gotta keep moving 😇 )

Thanks for your effort, and looking forward to the upgrade/updated version for storybook 6.4. Testing with storybook and testing-library looks very promising!

I'm afraid I'm probably not gonna be able to help you with the implementation details/problems. 😞 
But maybe using the typescript `ReturnType` might come in handy?

- https://twitter.com/NetanelBasal/status/1429819098133934080
- https://dev.to/busypeoples/notes-on-typescript-returntype-3m5a

**Comment by Marklb** [comment]
I never got around to fixing all the types errors on the previous versions, but imported story types should be fixed in `@storybook/testing-angular@1.0.3`. This hasn't had any activity in a while and I don't plan to go back and fix types on previous versions, so I will close this and assume this is fixed on the latest release.
